assembly/src/tokens/stream.rs

```diff
@@ -93,10 +114,55 @@ impl<'a> TokenStream<'a> {
             }
         }
     }
+
+    pub fn take_doc_comment_at(&mut self, pos: usize) -> Option<String> {
+        self.doc_comments.remove(&pos)?
+    }
 }
 
 impl<'a> fmt::Display for TokenStream<'a> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         write!(f, "{:?}", &self.tokens[self.pos..])
     }
 }
+
+#[derive(Debug)]
+pub struct Comment(Option<String>);
+
+impl Comment {
+    pub fn append_line(&mut self, line: &str) {
+        let prepared_line = Self::prepare_line(line);
+        if !prepared_line.is_empty() {
+            let mut prepared_line_string = String::from(prepared_line);
+            prepared_line_string.push('\n');
+            match &mut self.0 {
+                Some(comment) => {
+                    comment.push_str(&prepared_line_string);
+                }
+                None => self.0 = Some(prepared_line_string),
+            }
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.0.is_none()
+    }
+
+    pub fn take_content(&mut self) -> Option<String> {
+        match &mut self.0 {
+            Some(comment) => self.0 = Some(String::from(comment.trim())), // remove trailing \n
+            None => (),
+        }
```

Comment:
We shouldn't need this as it results in an extra allocation. (i.e., creating a new string could be expensive and we should avoid this). Instead of removing trailing `\n` here, we should construct the string in such a way that it doesn't end up with a trailing `\n`.